### PR TITLE
Update XNNPACK to latest version

### PR DESCRIPTION
**Pull request by skottmckay**
### Description

Update XNNPACK to latest version
- adds fp16 kernels and various other improvements
- requires pthreadpool update as well

Most code updates in the XNNPACK EP are to adjust to the new XNNPACK API
- 'setup' is split into 'reshape' and 'setup'
-  some ops use a workspace buffer
   -  copied workspace allocation from XNNPACK unit test code
- some suffixes changed 

Added wrapper for XNNPACK caches to base XNNPACK EP kernel
- simplifies usage
- XNNPACK split out the code and weights caches, but the code cache isn't currently usable via the public API
  - we could use the internal types if we think it's required for performance reasons. non-trivial though as we'd need to propagate ifdef values from the XNNPACK build up to the ORT build. 
  - using XNNPACK internals would also mean we would not be able to support using a pre-build XNNPACK package
    - not an issue currently
  
Fixed opset registration for internal NHWC domain
- was not being tied to the ONNX version, so nodes inserted by layout transformation had the incorrect opset
- a number of other places needed updating once this issue was fixed

Remove support for NCHW Resize from XNNPACK EP so it's NHWC only
- we only supported NCHW for fp32,
  - doing so adds complexity in multiple places (XNNPACK EP kernel implementation, layout transformation and transpose optimization)
  - unclear if that complexity provides any benefit. can add back if required by production scenario

### Motivation and Context

We're looking at enabling fp16 support for CoreML and NNAPI. If we do that we need a good fallback story if the CPU EP will be used. The XNNPACK fp16 kernels will hopefully provide that. 

NOTE: This PR doesn't add fp16 support to the XNNPACK EP kernels. That can be done as required in separate EPs and should be relatively simple to do.

### CODE REVIEW

File: onnxruntime/core/providers/xnnpack/nn/resize.cc [hunk 1]
```diff
@@ -18,6 +19,11 @@ bool Resize::IsOnnxNodeSupported(const NodeUnit& node_unit,
                                  const GraphViewer& graph_viewer) {
   bool supported = false;
   do {
+    // Internal NHWC domain starts at opset 11
```
Comment (wejoncy) [inline onnxruntime/core/providers/xnnpack/nn/resize.cc:22]: If the check is required for all layour sensitive ops, Could we put it ahead of `IsOnnxNodeSupported` in the entry function?

**Comment by skottmckay** [comment]
I'll address the deps.txt conflict once everyone is happy with the changes as it requires updating the dependencies package in the CI to match the latest `main`.

File: onnxruntime/core/providers/xnnpack/nn/average_pool.cc [hunk 2]
```diff
@@ -221,24 +227,49 @@ Status AveragePool::Compute(OpKernelContext* context) const {
...
+                           /*output_height_out=*/nullptr, /*output_width_out=*/nullptr,
+                           threadpool);
+
+  if (status != xnn_status_success) {
+    return ORT_MAKE_STATUS(ONNXRUNTIME, FAIL, "xnn_reshape_average_pooling2d_nhwc_", OpTypeToString(avgpool_type_),
+                           " returned ", status);
+  }
+
+  workspace.reset(allocator->aligned_allocate(allocator->context, workspace_size, XNN_ALLOCATION_ALIGNMENT));
+  status = xnn_setup_average_pooling2d_nhwc_f32(op0_.get(), workspace.get(),
```
Comment (wejoncy) [inline onnxruntime/core/providers/xnnpack/nn/average_pool.cc:255]: Do we need to call `xnn_setup_average_pooling2d_nhwc_f32` twice here?

File: onnxruntime/core/providers/xnnpack/nn/conv.cc [hunk 3]
```diff
@@ -64,29 +65,54 @@ Status Conv::Compute(OpKernelContext* context) const {
...
     return Status::OK();
   }
-  pthreadpool_t t_pool = GetThreadPool();
 
-  xnn_status status = xnn_status_invalid_state;
+  pthreadpool_t threadpool = GetThreadPool();
+
+  // setup allocator/automated dellocate for workspace
+  size_t workspace_size = 0;
+  size_t workspace_alignment = 0;
```
Comment (wejoncy) [inline onnxruntime/core/providers/xnnpack/nn/conv.cc:73]: Wondering if we supported workspace, are we gonna support subgraph as well?

Comment (skottmckay) [reply onnxruntime/core/providers/xnnpack/nn/conv.cc:73]: Based on the implementation where there were errors if the workspace wasn't provided, and the tests with the PR to update the API all used a workspace, it didn't appear to be optional.

e.g. https://github.com/google/XNNPACK/blob/c7e7cde37615a81a529c326aa278bfab4cd6fe5a/src/operators/average-pooling-nhwc.c#L1157

However I just noticed I didn't actually populate the workspace pointer in Conv::Compute so maybe it's not required. Didn't really make sense to me when we're allocating it for every Compute call anyway.

IIRC using subgraphs would completely change how we used XNNPACK from creating individual kernels to building subgraphs right? if using subgraphs is a non-trivial amount of work we'd want a solid use case to justify it.

File: onnxruntime/core/providers/xnnpack/nn/average_pool.cc (hunk 2 above)
Comment (skottmckay) [reply onnxruntime/core/providers/xnnpack/nn/average_pool.cc:255]: no - was accidental. good catch.

**Comment by wejoncy** [comment]
LGTM!

**Review by wejoncy — APPROVED** [review]

**Review by snnn — APPROVED** [review]

**Pull request merged**